src/froll.c

```diff
@@ -397,6 +397,277 @@ void frollsumExact(double *x, uint64_t nx, ans_t *ans, int k, double fill, bool
+          //Rprintf("iteration %d, new max %f at %d, old max %f at %d\n", i, x[i], i, w, iw);
+          iw = i;
+          w = x[iw];
+        } else {
+          if (iw > i-k) { // max is still within window
+            // w and iw are still ok
+            //Rprintf("iteration %d, max %f at %d bigger than current value %f\n", i, w, iw, x[i]);
+          } else { // max left the window, need to find max in this window
+            //Rprintf("iteration %d, max %f at %d left the window, call windowmax from %d of size %d\n", i, w, iw, i, k);
+            w = R_NegInf;
+            iw = i-k;
+            windowmax(x, i, k, &w, &iw);
+            //Rprintf("iteration %d, windowmax found new max %f at %d\n", i, w, iw);
+            cmax++;
+          }
+        }
+        ans->dbl_v[i] = w;
+      }
+      if (verbose)
+        snprintf(end(ans->message[0]), 500, _("%s: windowmax called %"PRIu64" time(s)\n"), __func__, cmax);
+      if (truehasna) {
+        if (hasna==-1) {
+          ans->status = 2;
+          snprintf(end(ans->message[2]), 500, _("%s: hasNA=FALSE used but NA (or other non-finite) value(s) are present in input, use default hasNA=NA to avoid this warning"), __func__);
+        }
+        if (verbose)
+          snprintf(end(ans->message[0]), 500, _("%s: NA (or other non-finite) value(s) are present in input, re-running with extra care for NAs\n"), __func__);
+        w = R_NegInf;
+        cmax = 0;
+      }
+    } else {
+      if (hasna==-1) {
+        ans->status = 2;
+        snprintf(end(ans->message[2]), 500, _("%s: hasNA=FALSE used but NA (or other non-finite) value(s) are present in input, use default hasNA=NA to avoid this warning"), __func__);
+      }
+      if (verbose)
+        snprintf(end(ans->message[0]), 500, _("%s: NA (or other non-finite) value(s) are present in input, skip non-NA attempt and run with extra care for NAs\n"), __func__);
+      w = R_NegInf;
+      cmax = 0;
+    }
+  }
+  if (truehasna) {
+    int nc = 0;
+    int i;
+    for (i=0; i<k-1; i++) { // #loop_counter_not_local_scope_ok
+      if (R_FINITE(x[i])) {
+        if (x[i] >= w) {
+          iw = i;
+          w = x[iw];
+        }
+      } else {
+        nc++;
+      }
+      ans->dbl_v[i] = fill;
+    }
+    if (R_FINITE(x[i])) {
+      if (x[i] >= w) {
+        iw = i;
+        w = x[iw];
+      }
+    } else {
+      nc++;
+    }
+    if (nc == 0) {
+      ans->dbl_v[i] = w;
+    } else if (nc == k) {
+      ans->dbl_v[i] = narm ? R_NegInf : NA_REAL;
+    } else {
+      ans->dbl_v[i] = narm ? w : NA_REAL;
+    }
+    for (uint64_t i=k; i<nx; i++) {
+      if (R_FINITE(x[i])) {
+        if (x[i] >= w) {
+          iw = i;
+          w = x[iw];
+        } else {
+          if (iw > i-k) { // max is still within window
+            // w and iw are still ok
+          } else { // max left the window, need to find max in this window
+            w = R_NegInf;
+            iw = i-k;
+            windowmaxnarm(x, i, k, narm, &nc, &w, &iw);
```

Comment:
here, maybe the worst case is something like

```
1 0 0 0 0 0 ... 1 0 0 .... 1 0 0 ...
```

Where `1` only shows up every `k+1` entries, so that we always have to double-pass the vector.

Again nothing actionable in this PR, just checking my own understanding+filing note for benchmarking.

Reply:
To draw more accurate conclusions it is probably better to look at final implementation in frollmax branch rather than frollmax1 as some parts have been simplified later on.

Reply:
Also manual explains edge case where naive approach (nested loop) algo=exact could ended up to be faster than algo=fast

Reply:
Thanks, noted. Resolving as nothing to do in this PR.